### PR TITLE
Improve printing of BoundsError with colon

### PR DESCRIPTION
**Pull request by nalimilan**
Avoid exposing implementation details like `Base.Slice(Base.OneTo(n))` and print `:` instead of `Colon()`.

Before:
```julia
julia> rand(2, 2)[1:10, :]
ERROR: BoundsError: attempt to access 2×2 Array{Float64,2} at index [1:10, Base.Slice(Base.OneTo(2))]
```

After:
```julia
julia> rand(2, 2)[1:10, :]
ERROR: BoundsError: attempt to access 2×2 Array{Float64,2} at index [1:10, 1:2]
```

A more "radical" approach would be to implement `print` for `Slice`, `OneTo` and `Colon` so that it avoids details, and call it from `showerror`. What do you think?

### CODE REVIEW

**Comment by bkamins** [comment]
Also I would consider checking in `showerror` for `BoundsError` if its `i` fields is an `AbstractString` and if so print it verbatim as passed instead of `"at index [some_index_here]"`. We already have a special case for `AbstractRange`. This would allow more flexibility in printing custom error messages that would be more informative for the users.

**Comment by JeffBezanson** [comment]
We could certainly use `:` for `show(:)`. `1:n` might also make sense as the `:compact` output of `OneTo(n)`?

**Comment by nalimilan** [comment]
> We could certainly use `:` for `show(:)`. `1:n` might also make sense as the `:compact` output of `OneTo(n)`?

I thought about `:compact`, but AFAIK it's only documented to be used to print fewer digits (`?IOContext` says "Boolean specifying that small values should be printed more compactly"). `?print` says "The representation used by print includes minimal formatting and tries to avoid Julia-specific details."  `:typeinfo` could also be used since the details we want to remove are all related to type information (but we would have to abuse it slightly by pretending the type is already available from `showerror`, which isn't the case).

**Comment by nalimilan** [comment]
Bump.

**Comment by StefanKarpinski** [comment]
Who should review, @nalimilan?

**Comment by bkamins** [comment]
@nalimilan apart from rebasing I think it would be good to test against all cases you define in `index_string`.

**Comment by nalimilan** [comment]
I've done that, plus a small improvement to avoid allocating a temporary string.

File: base/errorshow.jl
Comment (bkamins) [inline base/errorshow.jl:115]: how would that handle the case when `x.mask` is very long? Do we show then it compactly? (I guess we do but I just want to be sure)

Comment (nalimilan) [reply base/errorshow.jl:115]: Yes I get this:
```julia
julia> x[10, fill(true, 1000)]
ERROR: BoundsError: attempt to access 2×2 Array{Float64,2} at index [10, Bool[1, 1, 1, 1, 1, 1, 1, 1, 1, 1  …  1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]
```

**Review by bkamins — APPROVED** [review]

**Comment by nalimilan** [comment]
@JeffBezanson Good to merge?

**Comment by nalimilan** [comment]
I'll merge in a few days if nobody objects.

**Pull request merged**
